### PR TITLE
Move all schemadiff comparisons to canonical form

### PR DESCRIPTION
**Pull request by dbussink**
This changes all the internal schemadiff comparisons to use the canonical form to compare two parts for diffing.

The reason we need to do this is that there are inconsistencies in how things get canonicalized even in MySQL itself that we can't really fix in the parser either.

Take for example a functional index like the following:

```
UNIQUE KEY `deleted_check` (`id`, (if((`deleted_at` is null),0,NULL))))
```

Specifically, see the inconsistency with how `NULL` is treated. When it occurs as a singular value it's upcased, but in the `is null` it is always lowercased.

This gets very brittle in the parser to match the MySQL behavior 100% for the canonical form and I doubt it's even possible at all.

By doing diff comparisons with the schemadiff / Vitess canonical form, we at least are comparing these consistently. The Vitess parser will see this as `deleted_at` IS NULL but that is consistent then with itself.

When not doing this, we get stray `ALTER TABLE` computed for this since there's a casing difference but it does not matter.

## Related Issue(s)

Part of #10203

## Checklist

-   [ ] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/schemadiff/table.go [hunk 1]
```diff
@@ -1346,8 +1346,7 @@ func (c *CreateTableEntity) apply(diff *AlterTableEntityDiff) error {
 					return errors.Wrap(ErrApplyDuplicateKey, keyName)
 				}
 			}
-			for _, col := range opt.IndexDefinition.Columns {
-				colName := col.Column.String()
+			for colName := range getKeyColumnNames(opt.IndexDefinition) {
```
Comment (dbussink) [inline go/vt/schemadiff/table.go:1349]: This is a separate bug fix to ensure that we can apply an `ALTER` operation on a schema for a functional index. This uses the same logic as we use in the diff computation already.

File: go/vt/sqlparser/ast_format.go [hunk 2]
```diff
@@ -1673,7 +1673,7 @@ func (node *CreateView) Format(buf *TrackedBuffer) {
 		buf.literal("or replace ")
 	}
 	if node.Algorithm != "" {
-		buf.astPrintf(node, "algorithm = %s ", node.Algorithm)
+		buf.astPrintf(node, "algorithm = %#s ", node.Algorithm)
```
Comment (dbussink) [inline go/vt/sqlparser/ast_format.go:1676]: The algorithm should not always be upcased, so this treats that properly.

File: go/vt/schemadiff/schema.go [hunk 3]
```diff
@@ -234,7 +234,7 @@ func (s *Schema) Entities() []Entity {
 
 // EntityNames is a convenience function that returns just the names of entities, in good order
 func (s *Schema) EntityNames() []string {
-	names := []string{}
+	var names []string
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/schemadiff/schema.go:237]: Just saying, these two are not equivalent. I created an empty slice, and your IDE replaced that with a `nil` slice. Sometimes the difference is important.

**Review by shlomi-noach — APPROVED** [review]

File: go/vt/schemadiff/schema.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbussink) [reply go/vt/schemadiff/schema.go:237]: @shlomi-noach yeah, but we're already using nil slices as well for other methods like this one so I'm not that concerned here and prefer to be consistent with the other methods then.

**Pull request merged**